src/main/scala/viper/silver/plugin/standard/inline/InlineRewrite.scala

```diff
@@ -0,0 +1,79 @@
+package viper.silver.plugin.standard.inline
+
+import viper.silver.ast.{Exp, Fold, LocalVar, Method, PredicateAccess, Program, Stmt, Unfold}
+
+trait InlineRewrite {
+
+  def inlinePredicates(method: Method, program: Program): Method = {
+    val expandedPres = method.pres.map { pre =>
+      expandPredicate(pre, program).fold(pre)(expandedPred => expandedPred)
+    }
+    val expandedPosts = method.posts.map { post =>
+      expandPredicate(post, program).fold(post)(expandedPred => expandedPred)
+    }
+    method.copy(name = method.name,
+      formalArgs = method.formalArgs,
+      formalReturns = method.formalReturns,
+      pres = expandedPres,
+      posts = expandedPosts,
+      body = method.body
+    )(pos = method.pos, info = method.info, errT = method.errT)
+  }
+
+  def removeUnfoldFold(method: Method, program: Program): Method = {
+    val expandablePredicateIds = getExpandablePredicateIds(method.pres, program)
+    val rewrittenBody = method.body.map { body =>
+      val bodyWithRemovedUnfolds = removeUnfolds(body.ss, expandablePredicateIds)
+      val bodyWithRemovedUnfoldFolds = removeFolds(bodyWithRemovedUnfolds, expandablePredicateIds)
+      body.copy(ss = bodyWithRemovedUnfoldFolds, scopedDecls = body.scopedDecls)(
+        pos = body.pos, info = body.info, errT = body.errT
+      )
+    }
+    method.copy(name = method.name,
+      formalArgs = method.formalArgs,
+      formalReturns = method.formalReturns,
+      pres = method.pres,
+      posts = method.posts,
+      body = rewrittenBody
+    )(pos = method.pos, info = method.info, errT = method.errT)
+  }
+
+  private[this] def expandPredicate(expr: Exp, program: Program): Option[Exp] =
+    expr match {
+      case pred: PredicateAccess => getPredicateBody(pred, program)
+      case _ => None
+    }
+
+  private[this] def getPredicateBody(pred: PredicateAccess, program: Program): Option[Exp] = {
+    def extractLocalVarIds(predicate: PredicateAccess): Set[String] =
+      predicate.args.map {
+        case arg: LocalVar => arg.name
+        case _ => ""
+      }.filter(_.nonEmpty).toSet
+
+    val args = extractLocalVarIds(pred)
```

Comment:
`predicateBody` uses `args` in `Expressions.instantiateVariables`, which is then used as the scope in `Sanitizer.replaceFreeVariablesInExpression`, and this is what the doc says about that method:

```scala
  /** This method replaces free variables in the expression by corresponding subexpressions.
    * If bound variables exist and their identifiers clash with the new free variables or scope,
    * then the bound variables are renamed, preventing the new free variables from being 'captured'.
    ...
    * @param scope        A set of names representing the scope where the expression is located.
```

So `args` needs to be the current local variables in scope, not just the predicate arguments that happen to be local variables. I *think* this would be the same as the local variables in the scope of the method body, so

```scala
method.scopedDecls.map { _.name }.toSet
```

as I have in the comment above.